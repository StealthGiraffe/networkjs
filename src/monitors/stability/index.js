import { NetworkStatus } from '../../constants'
import { StabilityDefaults } from './Stability.constants'

export const ping = async resource => {
    try {
        const res = await fetch(resource)
        return res.status === 200
    } catch (e) {
        return false
    }
}

class StabilityMonitor {
    constructor(
        emitter,
        {
            resource,
            interval = StabilityDefaults.INTERVAL,
            requestThreshold = StabilityDefaults.REQUEST_THRESHOLD,
            durationThreshold = StabilityDefaults.DURATION_THRESHOLD,
        },
    ) {
        this.emitter = emitter
        this.resource = resource
        this.interval = interval
        this.requestThreshold = requestThreshold
        this.durationThreshold = durationThreshold

        this.initialize()
    }

    clearCurrentInterval() {
        if (this.runInterval) {
            clearInterval(this.runInterval)
            this.runInterval = null
        }
    }

    initialize() {
        this.consecutiveSlowRequestCount = 0
        this.paused = false
        this.clearCurrentInterval()
        this.runInterval = setInterval(this.run.bind(this), this.interval)
    }

    pause() {
        this.paused = true
        this.clearCurrentInterval()
    }

    resume() {
        this.initialize()
    }

    async run() {
        if (this.paused) return

        const start = window.performance.now()
        if (await ping(this.resource)) {
            if (window.performance.now() - start > this.durationThreshold) {
                this.consecutiveSlowRequestCount++
                if (
                    this.consecutiveSlowRequestCount ===
                        this.requestThreshold &&
                    !this.paused
                ) {
                    this.emitter.dispatchEvent(NetworkStatus.UNSTABLE)
                }
            } else {
                if (
                    this.consecutiveSlowRequestCount >= this.requestThreshold &&
                    !this.paused
                ) {
                    this.emitter.dispatchEvent(NetworkStatus.STABLE)
                }
                this.consecutiveSlowRequestCount = 0
            }
        }
    }
}

export default StabilityMonitor
