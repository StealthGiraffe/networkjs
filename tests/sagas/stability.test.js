// import { NetworkStatus } from '../constants'
import * as stability from '../../src/sagas/stability'
import EventEmitter from '../../src/events/EventEmitter'

const delay = stability.delay
const ping = stability.ping
const StabilitySaga = stability.default

describe('Stability Saga', () => {
    describe('delay', () => {
        it('waits the provided interval', () => {
            jest.useFakeTimers()
            delay(1000)

            expect(setTimeout).toHaveBeenCalledTimes(1)
            expect(setTimeout).toHaveBeenLastCalledWith(
                expect.any(Function),
                1000,
            )
        })
    })

    describe('ping', () => {
        it('returns true on fetch success', async () => {
            global.fetch = jest
                .fn()
                .mockImplementation(() =>
                    Promise.resolve({ ok: true, Id: '123' }),
                )
            const result = await ping('TEST')

            expect(global.fetch).toHaveBeenCalledTimes(1)
            expect(result).toBe(true)
        })

        it('returns false on fetch error', async () => {
            global.fetch = jest.fn().mockImplementation(() => Promise.reject())
            const result = await ping('TEST')

            expect(global.fetch).toHaveBeenCalledTimes(1)
            expect(result).toBe(false)
        })
    })

    describe('StabilitySaga', () => {
        let saga
        const emitter = new EventEmitter()

        beforeEach(() => {
            saga = new StabilitySaga(emitter, {
                resource: 'URL',
                interval: 5000,
                requestThreshold: 2000,
                durationThreshold: 2,
            })
        })

        describe('constructor', () => {
            it('initializes with the correct props', () => {
                expect(saga).toMatchSnapshot()
            })
        })

        describe('initialize', () => {
            it('sets some default properties and runs saga', () => {
                const runSpy = (StabilitySaga.prototype.run = jest.fn())
                saga.initialize()

                expect(saga.consecutiveSlowRequestCount).toBe(0)
                expect(saga.paused).toBe(false)
                expect(runSpy).toHaveBeenCalledTimes(1)
            })
        })

        describe('pause', () => {
            it('sets paused to true', () => {
                saga.pause()

                expect(saga.paused).toBe(true)
            })
        })

        describe('resume', () => {
            it('re-initializes saga', () => {
                const initializeSpy = (StabilitySaga.prototype.initialize = jest.fn())
                saga.resume()

                expect(saga.consecutiveSlowRequestCount).toBe(0)
                expect(saga.paused).toBe(false)
                expect(initializeSpy).toHaveBeenCalledTimes(1)
            })
        })

        describe('run', () => {
            it('pings the provided resource', async () => {
                const pingSpy = jest
                    .spyOn(stability, 'ping')
                    .mockImplementation(() => false)
                await saga.run()

                expect(pingSpy).toHaveBeenCalledTimes(1)
                expect(pingSpy).toHaveBeenLastCalledWith(URL)
            })
        })
    })
})

/*
class StabilitySaga {

    async run() {
        let shouldDelay = true
        const start = window.performance.now()
        const success = await ping(this.resource)
        if (success) {
            if (window.performance.now() - start > this.durationThreshold) {
                this.consecutiveSlowRequestCount++
                if (this.consecutiveSlowRequestCount < this.requestThreshold) {
                    shouldDelay = false
                } else if (
                    this.consecutiveSlowRequestCount ===
                        this.requestThreshold &&
                    !this.paused
                ) {
                    this.emitter.dispatchEvent(NetworkStatus.UNSTABLE)
                }
            } else {
                if (
                    this.consecutiveSlowRequestCount >= this.requestThreshold &&
                    !this.paused
                ) {
                    this.emitter.dispatchEvent(NetworkStatus.STABLE)
                }
                this.consecutiveSlowRequestCount = 0
            }
        }
        shouldDelay && (await delay(this.interval))
        !this.paused && (await this.run())
    }
}

export default StabilitySaga
*/
