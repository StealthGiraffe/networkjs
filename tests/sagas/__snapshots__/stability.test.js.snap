// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stability Saga StabilitySaga constructor initializes with the correct props 1`] = `
StabilitySaga {
  "consecutiveSlowRequestCount": 0,
  "durationThreshold": 2,
  "emitter": EventEmitter {
    "events": Object {},
  },
  "interval": 5000,
  "paused": false,
  "requestThreshold": 2000,
  "resource": "URL",
}
`;
